### PR TITLE
Protect DB directory with PID lock

### PR DESCRIPTION
**Pull request by mavam**
To ensure that only a single server process can write into a given DB directory, we now write a PID lock file when we start VAST.

### CODE REVIEW

File: libvast/src/detail/pid_file.cpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+      return contents.error();
+    return caf::make_error(ec::filesystem_error,
+                           "stale PID file found: ", filename.str(),
+                           "terminate process", *contents,
+                           "or remove PID file manually");
+  }
+  // Open the file.
+  auto fd = ::open(filename.str().c_str(), O_WRONLY | O_CREAT, 0600);
+  if (fd < 0)
+    return caf::make_error(ec::filesystem_error, "open(2):", strerror(errno));
```
Comment (tobim) [inline libvast/src/detail/pid_file.cpp:48]:
```suggestion
    return make_error(ec::filesystem_error, "open(2):", strerror(errno));
```

File: libvast/src/detail/pid_file.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  // Lock the file handle.
+  if (::flock(fd, LOCK_EX | LOCK_NB) < 0) {
+    ::close(fd);
+    return caf::make_error(ec::filesystem_error, "flock(2):", strerror(errno));
```
Comment (tobim) [inline libvast/src/detail/pid_file.cpp:52]:
```suggestion
    return make_error(ec::filesystem_error, "flock(2):", strerror(errno));
```

File: libvast/src/detail/pid_file.cpp [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  }
+  // Write the PID in human readable form into the file.
+  auto pid = to_string(process_id());
+  VAST_ASSERT(!pid.empty());
+  if (::write(fd, pid.data(), pid.size()) < 0) {
+    ::close(fd);
+    return caf::make_error(ec::filesystem_error, "write(2):", strerror(errno));
```
Comment (tobim) [inline libvast/src/detail/pid_file.cpp:59]:
```suggestion
    return make_error(ec::filesystem_error, "write(2):", strerror(errno));
```

File: libvast/src/detail/pid_file.cpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  // If the file exists, we cannot continue.
+  if (exists(filename)) {
+    // Attempt to read file to display an actionable error message.
+    auto contents = load_contents(filename);
+    if (!contents)
+      return contents.error();
+    return caf::make_error(ec::filesystem_error,
+                           "stale PID file found: ", filename.str(),
+                           "terminate process", *contents,
+                           "or remove PID file manually");
```
Comment (tobim) [inline libvast/src/detail/pid_file.cpp:43]:
```suggestion
    return make_error(ec::filesystem_error,
                      "stale PID file found: ", filename.str(),
                      "terminate process", *contents,
                      "or remove PID file manually");
```
Formatting might be wrong with this one.

File: libvast/src/detail/pid_file.cpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#include "vast/detail/pid_file.hpp"
+
+#include "vast/concept/parseable/numeric.hpp"
+#include "vast/concept/parseable/to.hpp"
```
Comment (tobim) [inline libvast/src/detail/pid_file.cpp:18]:
```suggestion
```

File: libvast/src/detail/pid_file.cpp (hunk 4 above)
Comment (mavam) [reply libvast/src/detail/pid_file.cpp:43]: Isn't the `make_error` overload going away with the next CAF release?

**Pull request merged**